### PR TITLE
Change logic to write xic.txt file

### PR DESCRIPTION
**Pull request by daducci**
When processing a tractogram with COMMIT, which contains let’s say 100 streamlines, we expect 100 coefficients as output. However, as some streamlines can be “pre-filtered” by `trk2dictionary` (e.g. too short), fewer coefficients (let’s say 98) may actually be estimated. To cope with that, at the moment we create a tractogram that matches this “smaller” set of streamlines, using the parameter `gen_trk` in `trk2dictionary`. This way, the coefficients that are saved to `xic.txt` with `save_results(save_coeff=True)` match this smaller tractogram (i.e. 98 coefficients for 98 streamlines).

However, this approach is not optimal because this breaks the logic, as the user expects one coefficient for each input streamline, so 100 and not 98; also, the creation of this additional tractogram occupies twice the amount of disk space.

This PR modifies COMMIT to save to the file `xic.txt` one coefficient for each input streamline (i.e. 100); the coefficients corresponding to the "pre-filtered" streamlines are set to 0. This also means that the parameter `gen_trk` of `trk2dictionary` is not necessary anymore (so it's removed).

In the case of models where multiple contributions are estimated for each streamline, e.g. `CylinderZeppelinBall`, then `xic` will contain all such coefficients, i.e. multiple values for every streamline.

### CODE REVIEW

File: commit/core.pyx [hunk 1]
```diff
@@ -740,7 +741,40 @@ cdef class Evaluation :
...
+            norm_fib = np.kron(np.ones(self.KERNELS['wmr'].shape[0]), self.DICTIONARY['TRK']['norm'])
+            x = self.x / np.hstack( (norm1*norm_fib,norm2,norm3) )
+        else :
+            x = self.x
+
+        offset1 = nF * self.KERNELS['wmr'].shape[0]
+        offset2 = offset1 + +nE*self.KERNELS['wmh'].shape[0]
+        kept = np.fromfile( pjoin(self.get_config('TRACKING_path'),'dictionary_TRK_kept.dict'), dtype=np.bool_)
+        xic = np.zeros( kept.shape )
+        xic[kept] = x[:offset1]
```
Comment (MarioOcampo) [inline commit/core.pyx:770]: This may be not compatible with `CylinderZeppelinBall` model

File: commit/core.pyx [hunk 2]
```diff
@@ -890,16 +923,17 @@ cdef class Evaluation :
             print( '\t\t- Coefficients txt files... ', end='' )
             sys.stdout.flush()
             if len(self.KERNELS['wmr']) > 0 :
+                kept = np.fromfile( pjoin(self.get_config('TRACKING_path'),'dictionary_TRK_kept.dict'), dtype=np.bool_)
+                xic = np.zeros( kept.shape )
                 offset = nF * self.KERNELS['wmr'].shape[0]
-                np.savetxt(pjoin(RESULTS_path,'xic.txt'), x[:offset], fmt='%12.5e')
+                xic[kept] = x[:offset]
```
Comment (MarioOcampo) [inline commit/core.pyx:929]: This may be not compatible with `CylinderZeppelinBall` model

File: commit/core.pyx (hunk 1 above)
Comment (daducci) [reply commit/core.pyx]: I made a modification, it should be fixed now

File: commit/core.pyx (hunk 2 above)
Comment (daducci) [reply commit/core.pyx]: Should be fixed now

**Comment by fullbat** [comment]
Instead of a flag, it might be used a parameter in order to decide how to save different coefficients for the case of multi-compartment models.
Something like: `save_coeff=average`  or `save_coeff=all`

**Comment by MarioOcampo** [comment]
Do you think the condition `if xec.size > 0`  should be checked to create the `compartment_EC.nii.gz`? The current implementation creates empty files if there is no `EC`.
(Same for `iso` compartment)

**Comment by daducci** [comment]
I don't mind having empty files, perhaps this way an analysis pipeline will not be broken, as a file always exist and the script should only check its length (same for the weights file)

File: commit/core.pyx [hunk 3]
```diff
@@ -879,28 +916,36 @@ cdef class Evaluation :
...
-            print( '\t\t- results.pickle... ', end='' )
-            sys.stdout.flush()
-            with open( pjoin(RESULTS_path,'results.pickle'), 'wb+' ) as fid :
-                pickle.dump( [self.CONFIG, self.x, x], fid, protocol=2 )
-            print( '[ OK ]' )
+        print( '\t\t- streamline_weights.txt... ', end='' )
+        sys.stdout.flush()
+        xic, _, _ = self.get_coeffs()
+        if stat_coeffs != 'all' and xic.size > 0 :
+            xic = np.reshape( xic, (-1,nF) )
```
Comment (MarioOcampo) [inline commit/core.pyx:923]: This is the previous idea, 
the size of `dictionary_TRK_kept` should be used instead of `nF`

Comment (daducci) [reply commit/core.pyx]: You're right! Now this should be fixed

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,16 @@
 # Change Log
 All notable changes to COMMIT will be documented in this file.
 
+## [1.4.0] - 2020-06-19
```
Comment (MarioOcampo) [inline CHANGELOG.md:5]: Maybe update the date?

File: setup.py [hunk 5]
```diff
@@ -44,7 +44,7 @@ def run(self):
 
 description = 'Convex Optimization Modeling for Microstructure Informed Tractography (COMMIT)'
 opts = dict(name='dmri-commit',
-            version='1.3.9.2',
+            version='1.4.0.0',
```
Comment (MarioOcampo) [inline setup.py:47]: The numbering format in `CHANGELOG.md` is different

Comment (daducci) [reply setup.py:47]: Done!

**Pull request merged**